src/security/core/tests/common/cryptography_wrapper.c

```diff
@@ -109,6 +125,219 @@ void set_entity_data_secret(struct dds_security_cryptography_impl * impl, const
+  case REMOTE_WRITER_TOKENS: return "REMOTE_WRITER_TOKENS";
+  case REMOTE_READER_TOKENS: return "REMOTE_READER_TOKENS";
+  default: assert (0); return "";
+  }
+}
+
+static void print_tokens (enum crypto_tokens_type type, const DDS_Security_ParticipantCryptoHandle lch, const DDS_Security_ParticipantCryptoHandle rch,
+    const DDS_Security_ParticipantCryptoTokenSeq *tokens)
+{
+  ddsrt_mutex_lock (&g_print_token_lock);
+  printf ("Token type %s, local %"PRIx64" / remote %"PRIx64", count: %u\n", get_crypto_token_type_str (type), lch, rch, tokens->_length);
+  for (uint32_t i = 0; i < tokens->_length; i++)
+  {
+    printf ("- token: ");
+    for (uint32_t j = 0; j < tokens->_buffer[i].binary_properties._buffer[0].value._length && j < 32; j++)
+      printf ("%02x", tokens->_buffer[i].binary_properties._buffer[0].value._buffer[j]);
+    printf ("\n");
+  }
+  printf ("\n");
+  ddsrt_mutex_unlock (&g_print_token_lock);
+}
+
+static void add_tokens (struct ddsrt_circlist *list, enum crypto_tokens_type type,
+    const DDS_Security_ParticipantCryptoHandle lch, const DDS_Security_ParticipantCryptoHandle rch,
+    const DDS_Security_ParticipantCryptoTokenSeq *tokens)
+{
+  struct crypto_token_data *token_data = ddsrt_malloc (sizeof (*token_data));
+  token_data->type = type;
+  token_data->local_handle = lch;
+  token_data->remote_handle = rch;
+  token_data->n_tokens = tokens->_length;
+  assert (tokens->_length <= CRYPTO_TOKEN_MAXLEN);
+  for (uint32_t i = 0; i < tokens->_length; i++)
+  {
+    size_t len = tokens->_buffer[i].binary_properties._buffer[0].value._length;
+    assert (len <= CRYPTO_TOKEN_SIZE);
+    memcpy (token_data->data[i], tokens->_buffer[i].binary_properties._buffer[0].value._buffer, len);
+    token_data->data_len[i] = len;
+  }
+  ddsrt_circlist_append(list, &token_data->e);
+}
+
+static void store_tokens (struct dds_security_crypto_key_exchange_impl *impl, enum crypto_tokens_type type, const DDS_Security_ParticipantCryptoHandle lch,
+    const DDS_Security_ParticipantCryptoHandle rch, const DDS_Security_ParticipantCryptoTokenSeq *tokens)
+{
+  if (!check_crypto_tokens ((const DDS_Security_DataHolderSeq *) tokens))
+  {
+    printf ("%d ERROR\n", type);
+    return;
+  }
+
+  ddsrt_mutex_lock (&impl->parent->token_data_lock);
+  add_tokens (&impl->parent->token_data_list, type, lch, rch, tokens);
+  ddsrt_mutex_unlock (&impl->parent->token_data_lock);
+
+  print_tokens (type, lch, rch, tokens);
+}
+
+struct ddsrt_circlist * get_crypto_tokens (struct dds_security_cryptography_impl * impl)
+{
+  struct ddsrt_circlist *tokens = ddsrt_malloc (sizeof (*tokens));
+  ddsrt_circlist_init (tokens);
+
+  ddsrt_mutex_lock (&impl->token_data_lock);
+  if (ddsrt_circlist_isempty (&impl->encode_decode_log))
+  {
+    ddsrt_mutex_unlock (&impl->token_data_lock);
+    return tokens;
+  }
+
+  struct ddsrt_circlist_elem *elem0 = ddsrt_circlist_oldest (&impl->token_data_list), *elem = elem0;
+  while (elem != NULL)
```

Comment:
`elem` can't be a null pointer here, I'd just a `do { ... } while (elem != elem0);`